examples/run_CPMG.py

```diff
@@ -4,149 +4,184 @@
-    nPhaseSteps = 1
+# NOTE: The overall phase and the 90-180 phase difference are phase cycled
+# in a nested way
+ph2 = r_[0, 1, 2, 3]
+ph_diff = r_[0, 2]
+ph1_cyc = array([(j + k) % 4 for k in ph2 for j in ph_diff])
+ph2_cyc = array([(k + 1) % 4 for k in ph2 for j in ph_diff])
+nPhaseSteps = len(ph2) * len(ph_diff)
 # }}}
-# {{{better tau
-marker = 1.0  # 10/10/22 → what is this? → pretty sure the time needed to execute the marker command
-pad_start = config_dict["tau_extra_us"] - config_dict["deadtime_us"]
-pad_end = (
-    config_dict["tau_extra_us"] - config_dict["deblank_us"] - marker
-)  # marker + deblank
-assert(pad_start > 0), "tau_extra_us must be set to more than deadtime and more than deblank!"
-assert (pad_end > 0), "tau_extra_us must be set to more than deadtime and more than deblank!"
-twice_tau_echo_us = (  # the period between 180 pulses
-    config_dict["tau_extra_us"] * 2 + config_dict["acq_time_ms"] * 1e3
+# {{{ calibrate pulse lengths
+# NOTE: This is done inside the run_spin_echo rather than in the example
+# but to keep the generic function more robust we do it outside of the ppg
+prog_p90_us = prog_plen(config_dict["p90_us"])
+prog_p180_us = prog_plen(2 * config_dict["p90_us"])
+# }}}
+# {{{ calculate symmetric tau
+# NOTE: here the tau_us is defined as the evolution time from the start of
+# excitation (*during the pulse*) through to the start of the 180 pulse
+marker_us = 1.0  # the marker takes 1 us
+config_dict["tau_us"] = (
+    2 * config_dict["deadtime_us"] + 1e3 * config_dict["echo_acq_ms"]
+) / 2
+assert (
+    config_dict["tau_us"] > 2 * prog_p90_us / pi + marker_us + config_dict["deblank_us"]
 )
-# now twice_tau_echo_us/2.0 is τ_echo, so I need to subtract the extra delays
-# imposed by the ppg to determine the remaining τ that I need
-config_dict["tau_us"] = twice_tau_echo_us / 2.0 - (
-    2
-    * config_dict["p90_us"]
-    / pi  # evolution during pulse -- see eq 6 of coherence paper
-    + config_dict["deadtime_us"]  # following 90
-    + config_dict["deblank_us"]  # before 180
+assert config_dict["deadtime_us"] > config_dict["deblank_us"] + 2 * marker_us
+print(
+    "If you are measuring on a scope, the time from the start (or end) of one 180 pulse to the next should be %0.1f us"
+    % (2 * config_dict["deadtime_us"] + 1e3 * config_dict["echo_acq_ms"] + prog_p180_us)
 )
 # }}}
 # {{{check total points
-total_pts = nPoints * nPhaseSteps
-assert total_pts < 2 ** 14, (
-    "You are trying to acquire %d points (too many points) -- either change SW or acq time so nPoints x nPhaseSteps is less than 16384\nyou could try reducing the acq_time_ms to %f"
-    % (total_pts, config_dict["acq_time_ms"] * 16384 / total_pts)
+total_pts = nPoints * nPhaseSteps * config_dict["nEchoes"]
+assert total_pts < 2**14, (
+    "You are trying to acquire %d points (too many points) -- either change SW or acq time so nPoints x nPhaseSteps is less than 16384"
+    % total_pts
 )
 # }}}
-# {{{run cpmg
+# {{{ acquire CPMG
```

Comment:
more in line with how run_hahn_echo.py and run_generic_echo.py label this section